Adding new algorithm for Longest Increasing Subsequence.

hijajoo opened the pull request:
<!-- Enter a brief description of the changes you've made in the next line -->
Adding algorithm and testcases for Longest Increasing Subsequence in folder dynamic_programming.
<!-- Check the following boxes, if applicable, by replacing the space inside
	"[ ]" with an "x", eg. [x] -->
- [X] Have read the [contribution guidelines][contrib-guidelines]
- [X] Closes an issue: close #153  <!-- enter the issue number right after the #, eg. #123-->
- [X] Includes unit tests (or unit tests have already been added)

[contrib-guidelines]: https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md

hj268 committed:
Adding new algorithm for Longest Increasing Subsequence.
Adding 5 testcases for Longest Increasing Subsequence.

faheel commented on C++/source/dynamic_programming/longest_increasing_subsequence.cpp:
```diff
@@ -0,0 +1,96 @@
+
+#ifndef LONGEST_INCREASING_SUBSEQUENCE_TEST
+int main() {
+    return 0;
```
An empty `main()` is not a good idea. Can you please add some code to take input from the user, call `longest_increasing_subsequence()`, and display it's result too?

faheel commented on C++/test/dynamic_programming/longest_increasing_subsequence.test.cpp:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+}
+
```
Please add 2 more test cases, one in which some numbers are negative, and one in which all numbers are negative in the expected longest increasing subsequence.

faheel commented on C++/source/dynamic_programming/longest_increasing_subsequence.cpp:
```diff
@@ -0,0 +1,96 @@
+vector<int> longest_increasing_subsequence(const vector<int> &a) {
+    vector<int> lis_count(a.size());
+    vector<int> lis_prev(a.size());
+    vector<int> ans;
```
Declare variables just before they're used for the first time (which would be line 80 in case of `ans`), and not at the very top of a function.

faheel commented on C++/source/dynamic_programming/longest_increasing_subsequence.cpp:
```diff
@@ -0,0 +1,96 @@
+    return -1;
+}
+
+int index_of_max(const vector<int> &a) {
```
Add a comment that briefly describes what this function does.
Also, consider a better name for `a` (`values`? `input`?).

faheel commented on C++/source/dynamic_programming/longest_increasing_subsequence.cpp:
```diff
@@ -0,0 +1,96 @@
+
+using namespace std;
+
+int find_prev_smaller(const vector<int> &a, int x) {
```
Add a comment that briefly describes what this function does.
Also, consider a better name for `a` (`values`? `input`?) and `x` (`current_index`?).

faheel reviewed (changes requested):
Thank you for contributing! There are a few changes that I think would be required before merging the PR.

faheel commented on C++/source/dynamic_programming/longest_increasing_subsequence.cpp:
```diff
@@ -0,0 +1,96 @@
+
+   Time complexity
+   --------------
+   O(N), where N is size of array
```
I think it's *O*(N<sup>2</sup>) since for every index you call `find_previous_smaller()`, which takes linear time. So for `N` number of indices, it would take *O*(N<sup>2</sup>) time.

hj268 committed:
Addressing review comments.
1. Added comments describing helper functions
2. Changed names of parameter variables.
3. Changed the comment describing time complexity.
4. Added 2 more testcases with negative numbers.
5. In main function, added code to call LIS on user inputed array.

hj268 committed:
Fixing a GCC warning

hijajoo commented:
Travis CI errors seem to be unrelated to my changes.

faheel commented on C++/test/dynamic_programming/longest_increasing_subsequence.test.cpp:
```diff
@@ -0,0 +1,104 @@
+        REQUIRE(lis.size() == expected_lis.size());
+    }
+    for(int i = 0; i < (int)lis.size(); i++) {
+        SECTION("Checking value at index " + i) {
```
`SECTION` block is not required here (and in other test cases too).

faheel commented:
@hijajoo I found a test case that gives wrong output:
Input: `2 6 4 8 7 -7 5 2 1 4 6 8 9 0 2 3` (size 16)
Expected output: `-7 1 4 6 8 9`
Actual output: `-7 1 4 6 8 9 0 2 3`

alxmjo commented:
Looks like this hasn't been updated in a while. Closing.
